File: web/src/game/combat/fighter.ts
import { Action } from "./actions";
import * as C from "./constants";
import type { AttackDef } from "./constants";
import type { CharacterDef } from "../data/characters";

export enum FighterState {
  IDLE = 0,
  WALK = 1,
  JUMP = 2,
  LIGHT = 3,
  HEAVY = 4,
  BLOCK = 5,
  HITSTUN = 6,
  KO = 7,
  BLOCKSTUN = 8,
  GUARD_BREAK = 9,
  PARRY = 10,
  PARRY_WHIFF = 11,
}

export type Rect = { left: number; top: number; right: number; bottom: number };

function scaleAtk(base: AttackDef, char: CharacterDef): AttackDef {
  return {
    ...base,
    damage: base.damage * char.stats.damage,
    chip: base.chip * char.stats.damage,
    staminaCost: base.staminaCost * char.stats.staminaCost,
    knockback: base.knockback * char.stats.knockback,
  };
}

export class Fighter {
  index: number;
  characterId: string;
  x: number;
  y = C.GROUND_Y;
  vx = 0;
  vy = 0;
  facing = 1;
  hp = C.MAX_HP;
  maxHp = C.MAX_HP;
  stamina = C.MAX_STAMINA;
  maxStamina = C.MAX_STAMINA;
  state: FighterState = FighterState.IDLE;
  stateTimer = 0;
  attackTimer = 0;
  attackDef: AttackDef | null = null;
  hitConnected = false;
  onGround = true;
  lastAction: Action = Action.IDLE;
  blockHoldFrames = 0;
  staminaRegenCd = 0;
  justParried = false;
  justGuardBroke = false;
  justParryWhiffed = false;
  parryConnected = false;
  lastParryDamageTaken = 0;

  // character-scaled kits
  private light: AttackDef;
  private heavy: AttackDef;
  private walkSpeed: number;
  private jumpVel: number;

  constructor(index: number, x: number, facing: number, character: CharacterDef) {
    this.index = index;
    this.characterId = character.id;
    this.x = x;
    this.facing = facing;
    this.maxHp = C.MAX_HP * character.stats.hp;
    this.hp = this.maxHp;
    this.maxStamina = C.MAX_STAMINA * character.stats.stamina;
    this.stamina = this.maxStamina;
    this.light = scaleAtk(C.LIGHT, character);
    this.heavy = scaleAtk(C.HEAVY, character);
    this.walkSpeed = C.WALK_SPEED * character.stats.speed;
    this.jumpVel = C.JUMP_VELOCITY * character.stats.jump;
  }

  get alive() {
    return this.hp > 0 && this.state !== FighterState.KO;
  }

  get locked() {
    return (
      this.state === FighterState.LIGHT ||
      this.state === FighterState.HEAVY ||
      this.state === FighterState.HITSTUN ||
      this.state === FighterState.BLOCKSTUN ||
      this.state === FighterState.GUARD_BREAK ||
      this.state === FighterState.PARRY ||
      this.state === FighterState.PARRY_WHIFF ||
      this.state === FighterState.KO
    );
  }

  get isAttacking() {
    return this.state === FighterState.LIGHT || this.state === FighterState.HEAVY;
  }

  get isShielding() {
    return this.state === FighterState.BLOCK && this.onGround;
  }

  get inParryWindow() {
    return (
      this.isShielding &&
      this.blockHoldFrames > 0 &&
      this.blockHoldFrames <= C.PARRY_WINDOW_FRAMES
    );
  }

  get attackActive() {
    if (!this.isAttacking || !this.attackDef) return false;
    const ad = this.attackDef;
    return this.attackTimer >= ad.startup && this.attackTimer < ad.startup + ad.active;
  }

  bodyRect(): Rect {
    const half = C.FIGHTER_WIDTH * 0.5;
    return {
      left: this.x - half,
      top: this.y - C.FIGHTER_HEIGHT,
      right: this.x + half,
      bottom: this.y,
    };
  }

  shieldRect(): Rect | null {
    if (!this.isShielding && this.state !== FighterState.BLOCKSTUN) return null;
    const halfH = C.FIGHTER_HEIGHT * 0.55;
    const cx = this.x + this.facing * (C.FIGHTER_WIDTH * 0.55);
    const top = this.y - C.FIGHTER_HEIGHT * 0.85;
    const w = 14;
    return this.facing >= 0
      ? { left: cx, top, right: cx + w, bottom: top + halfH }
      : { left: cx - w, top, right: cx, bottom: top + halfH };
  }

  hitboxRect(): Rect | null {
    if (!this.attackActive || !this.attackDef) return null;
    const ad = this.attackDef;
    const left = this.facing >= 0 ? this.x : this.x - ad.range;
    const right = this.facing >= 0 ? this.x + ad.range : this.x;
    const top = this.y + ad.hitboxYOffset - ad.hitboxH * 0.5;
    return { left, top, right, bottom: top + ad.hitboxH };
  }

  canAfford(cost: number) {
    return this.stamina >= cost;
  }

  spendStamina(amount: number) {
    this.stamina = Math.max(0, this.stamina - amount);
    this.staminaRegenCd = C.STAMINA_REGEN_DELAY;
  }

  applyAction(action: Action) {
    this.lastAction = action;
    this.justParried = false;
    this.justGuardBroke = false;
    this.justParryWhiffed = false;

    if (this.state === FighterState.KO) return;
    if (
      this.state === FighterState.HITSTUN ||
      this.state === FighterState.BLOCKSTUN ||
      this.state === FighterState.GUARD_BREAK ||
      this.state === FighterState.PARRY ||
      this.state === FighterState.PARRY_WHIFF
    ) {
      return;
    }
    if (this.isAttacking) return;

    if (action === Action.BLOCK && this.onGround && this.stamina > 0) {
      if (this.state !== FighterState.BLOCK) {
        this.blockHoldFrames = 0;
        this.parryConnected = false;
      }
      this.state = FighterState.BLOCK;
      this.blockHoldFrames += 1;
      this.vx = 0;
      this.spendStamina(C.BLOCK_HOLD_DRAIN);
      if (this.stamina <= 0) {
        this.stamina = 0;
        this.releaseBlock(true);
      }
      return;
    }

    if (this.state === FighterState.BLOCK) {
      this.releaseBlock(true);
      // releaseBlock may enter PARRY_WHIFF recovery
      if (this.locked) return;
    }

    if (action === Action.LIGHT && this.onGround) {
      if (this.canAfford(this.light.staminaCost)) {
        this.spendStamina(this.light.staminaCost);
        this.startAttack(this.light, FighterState.LIGHT);
      }
      return;
    }

    if (action === Action.HEAVY && this.onGround) {
      if (this.canAfford(this.heavy.staminaCost)) {
        this.spendStamina(this.heavy.staminaCost);
        this.startAttack(this.heavy, FighterState.HEAVY);
      }
      return;
    }

    if (action === Action.JUMP && this.onGround) {
      if (this.canAfford(C.JUMP_STAMINA_COST)) {
        this.spendStamina(C.JUMP_STAMINA_COST);
        this.vy = this.jumpVel;
        this.onGround = false;
        this.state = FighterState.JUMP;
      }
      return;
    }

    const moving = action === Action.LEFT || action === Action.RIGHT;
    if (moving) {
      const drain = this.onGround
        ? C.WALK_STAMINA_PER_FRAME
        : C.AIR_MOVE_STAMINA_PER_FRAME;
      if (this.stamina < C.MOVE_MIN_STAMINA) {
        if (this.onGround) {
          this.vx = 0;
          this.state = FighterState.IDLE;
        }
        return;
      }
      this.spendStamina(drain);
      const speed = this.onGround ? this.walkSpeed : this.walkSpeed * C.AIR_CONTROL;
      this.vx = action === Action.LEFT ? -speed : speed;
      if (this.onGround) this.state = FighterState.WALK;
      return;
    }

    if (this.onGround) {
      this.vx = 0;
      this.state = FighterState.IDLE;
    }
  }

  private releaseBlock(whiffCheck: boolean) {
    const held = this.blockHoldFrames;
    this.blockHoldFrames = 0;
    if (
      whiffCheck &&
      !this.parryConnected &&
      held > 0 &&
      held <= C.PARRY_WINDOW_FRAMES
    ) {
      this.justParryWhiffed = true;
      this.state = FighterState.PARRY_WHIFF;
      this.stateTimer = C.PARRY_WHIFF_RECOVERY;
      this.spendStamina(C.PARRY_WHIFF_STAMINA_COST);
      this.vx = 0;
    } else {
      this.state = FighterState.IDLE;
    }
    this.parryConnected = false;
  }

  private startAttack(attack: AttackDef, state: FighterState) {
    this.state = state;
    this.attackDef = attack;
    this.attackTimer = 0;
    this.hitConnected = false;
    this.vx = 0;
    this.blockHoldFrames = 0;
  }

  tickTimersAndPhysics() {
    this.regenStamina();

    if (this.state === FighterState.KO) {
      this.vx *= 0.9;
      this.integrate();
      return;
    }

    if (
      this.state === FighterState.HITSTUN ||
      this.state === FighterState.BLOCKSTUN ||
      this.state === FighterState.GUARD_BREAK ||
      this.state === FighterState.PARRY ||
      this.state === FighterState.PARRY_WHIFF
    ) {
      this.stateTimer -= 1;
      this.vx *= 0.85;
      if (this.stateTimer <= 0) {
        this.state = this.onGround ? FighterState.IDLE : FighterState.JUMP;
        this.vx = 0;
        this.blockHoldFrames = 0;
      }
      this.integrate();
      return;
    }

    if (this.isAttacking && this.attackDef) {
      this.attackTimer += 1;
      const total =
        this.attackDef.startup + this.attackDef.active + this.attackDef.recovery;
      if (this.attackTimer >= total) {
        this.state = this.onGround ? FighterState.IDLE : FighterState.JUMP;
        this.attackDef = null;
        this.attackTimer = 0;
        this.hitConnected = false;
      }
      this.integrate();
      return;
    }

    if (this.state === FighterState.BLOCK) this.vx = 0;
    this.integrate();
    if (this.onGround && this.state === FighterState.JUMP) this.state = FighterState.IDLE;
  }

  private regenStamina() {
    if (this.staminaRegenCd > 0) {
      this.staminaRegenCd -= 1;
      return;
    }
    if (
      this.state === FighterState.LIGHT ||
      this.state === FighterState.HEAVY ||
      this.state === FighterState.BLOCK ||
      this.state === FighterState.BLOCKSTUN ||
      this.state === FighterState.GUARD_BREAK ||
      this.state === FighterState.PARRY_WHIFF ||
      this.state === FighterState.WALK ||
      this.state === FighterState.KO
    ) {
      return;
    }
    this.stamina = Math.min(this.maxStamina, this.stamina + C.STAMINA_REGEN_PER_FRAME);
  }

  private integrate() {
    if (!this.onGround || this.vy < 0) {
      this.vy = Math.min(this.vy + C.GRAVITY, C.MAX_FALL_SPEED);
    }
    this.x += this.vx;
    this.y += this.vy;
    this.x = Math.max(C.WALL_MARGIN, Math.min(C.STAGE_WIDTH - C.WALL_MARGIN, this.x));
    if (this.y >= C.GROUND_Y) {
      this.y = C.GROUND_Y;
      this.vy = 0;
      this.onGround = true;
    } else {
      this.onGround = false;
    }
  }

  facingAttacker(attackerX: number) {
    return attackerX >= this.x ? this.facing > 0 : this.facing < 0;
  }

  takeHit(opts: {
    damage: number;
    hitstun: number;
    knockback: number;
    direction: number;
    attackerX: number;
    chip: number;
    blockStaminaCost: number;
    attackBlockstun: number;
  }): { dealt: number; result: string } {
    if (this.state === FighterState.KO) return { dealt: 0, result: "none" };
    if (this.state === FighterState.PARRY) return { dealt: 0, result: "none" };

    const canDefend =
      this.state === FighterState.BLOCK &&
      this.onGround &&
      this.facingAttacker(opts.attackerX) &&
      this.stamina > 0;

    if (canDefend && this.inParryWindow) {
      this.justParried = true;
      this.parryConnected = true;
      this.state = FighterState.PARRY;
      this.stateTimer = C.PARRY_FLASH_FRAMES;
      this.blockHoldFrames = 0;
      this.attackDef = null;
      this.attackTimer = 0;
      this.stamina = Math.min(this.maxStamina, this.stamina + C.PARRY_STAMINA_REFUND);
      this.vx = 0;
      return { dealt: 0, result: "parry" };
    }

    if (canDefend) {
      this.spendStamina(opts.blockStaminaCost);
      if (this.stamina <= 0) {
        this.stamina = 0;
        this.justGuardBroke = true;
        const dealt = opts.damage;
        this.hp = Math.max(0, this.hp - dealt);
        this.blockHoldFrames = 0;
        this.attackDef = null;
        this.attackTimer = 0;
        if (this.hp <= 0) {
          this.ko();
          this.vx = opts.direction * opts.knockback;
          return { dealt, result: "guard_break" };
        }
        this.state = FighterState.GUARD_BREAK;
        this.stateTimer = C.GUARD_BREAK_STUN;
        this.vx = opts.direction * opts.knockback * 1.1;
        return { dealt, result: "guard_break" };
      }
      const dealt = opts.chip;
      this.hp = Math.max(0, this.hp - dealt);
      this.state = FighterState.BLOCKSTUN;
      this.stateTimer = opts.attackBlockstun;
      this.blockHoldFrames = 0;
      this.vx = opts.direction * opts.knockback * C.BLOCK_PUSHBACK_SCALE;
      if (this.hp <= 0) this.ko();
      return { dealt, result: "block" };
    }

    const dealt = opts.damage;
    this.hp = Math.max(0, this.hp - dealt);
    this.blockHoldFrames = 0;
    this.attackDef = null;
    this.attackTimer = 0;
    if (this.hp <= 0) {
      this.ko();
      this.vx = opts.direction * opts.knockback;
      return { dealt, result: "hit" };
    }
    this.state = FighterState.HITSTUN;
    this.stateTimer = opts.hitstun;
    this.vx = opts.direction * opts.knockback;
    return { dealt, result: "hit" };
  }

  applyParryPunish(direction: number, isHeavy: boolean): number {
    if (this.state === FighterState.KO) return 0;
    const dmg =
      C.PARRY_COUNTER_DAMAGE + (isHeavy ? C.PARRY_COUNTER_HEAVY_BONUS : 0);
    this.hp = Math.max(0, this.hp - dmg);
    this.attackDef = null;
    this.attackTimer = 0;
    this.hitConnected = true;
    this.blockHoldFrames = 0;
    this.vx = direction * C.PARRY_COUNTER_KNOCKBACK;
    this.lastParryDamageTaken = dmg;
    if (this.hp <= 0) {
      this.ko();
      return dmg;
    }
    this.state = FighterState.HITSTUN;
    this.stateTimer = C.PARRY_STUN_ATTACKER;
    return dmg;
  }

  private ko() {
    this.hp = 0;
    this.state = FighterState.KO;
    this.stateTimer = 0;
    this.attackDef = null;
    this.blockHoldFrames = 0;
    this.parryConnected = false;
  }
}

export function rectsOverlap(a: Rect, b: Rect) {
  return a.left < b.right && a.right > b.left && a.top < b.bottom && a.bottom > b.top;
}

File: bit_duel/core/fighter.py
"""Fighter state machine + physics. Pure Python, no rendering."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import Optional

from . import constants as C
from .actions import Action


class FighterState(IntEnum):
    IDLE = 0
    WALK = 1
    JUMP = 2
    LIGHT = 3
    HEAVY = 4
    BLOCK = 5
    HITSTUN = 6
    KO = 7
    BLOCKSTUN = 8  # locked after absorbing a hit on shield
    GUARD_BREAK = 9  # stamina emptied while blocking
    PARRY = 10  # brief success state (invuln / flash)
    PARRY_WHIFF = 11  # failed parry attempt — recovery, fully open


@dataclass
class AttackDef:
    startup: int
    active: int
    recovery: int
    damage: float
    chip: float
    hitstun: int
    blockstun: int
    knockback: float
    range: float
    hitbox_h: float
    hitbox_y_offset: float
    stamina_cost: float
    is_heavy: bool = False

    @property
    def total_frames(self) -> int:
        return self.startup + self.active + self.recovery


LIGHT_ATK = AttackDef(
    startup=C.LIGHT_STARTUP,
    active=C.LIGHT_ACTIVE,
    recovery=C.LIGHT_RECOVERY,
    damage=C.LIGHT_DAMAGE,
    chip=C.LIGHT_CHIP,
    hitstun=C.LIGHT_HITSTUN,
    blockstun=C.LIGHT_BLOCKSTUN,
    knockback=C.LIGHT_KNOCKBACK,
    range=C.LIGHT_RANGE,
    hitbox_h=C.LIGHT_HITBOX_H,
    hitbox_y_offset=C.LIGHT_HITBOX_Y_OFFSET,
    stamina_cost=C.LIGHT_STAMINA_COST,
    is_heavy=False,
)

HEAVY_ATK = AttackDef(
    startup=C.HEAVY_STARTUP,
    active=C.HEAVY_ACTIVE,
    recovery=C.HEAVY_RECOVERY,
    damage=C.HEAVY_DAMAGE,
    chip=C.HEAVY_CHIP,
    hitstun=C.HEAVY_HITSTUN,
    blockstun=C.HEAVY_BLOCKSTUN,
    knockback=C.HEAVY_KNOCKBACK,
    range=C.HEAVY_RANGE,
    hitbox_h=C.HEAVY_HITBOX_H,
    hitbox_y_offset=C.HEAVY_HITBOX_Y_OFFSET,
    stamina_cost=C.HEAVY_STAMINA_COST,
    is_heavy=True,
)


@dataclass
class Fighter:
    """One combatant. Index 0 is usually P1 (left), 1 is P2 (right)."""

    index: int
    x: float
    y: float = C.GROUND_Y
    vx: float = 0.0
    vy: float = 0.0
    facing: int = 1  # +1 right, -1 left
    hp: float = C.MAX_HP
    stamina: float = C.MAX_STAMINA
    state: FighterState = FighterState.IDLE
    state_timer: int = 0  # frames remaining in locked states
    attack_timer: int = 0  # frames into current attack
    attack_def: Optional[AttackDef] = None
    hit_connected: bool = False
    on_ground: bool = True
    last_action: Action = Action.IDLE
    # Shield: frames spent holding block on this press (0 = not blocking)
    block_hold_frames: int = 0
    # Frames until stamina regen resumes after a spend
    stamina_regen_cd: int = 0
    # Set for one frame after a successful parry (match uses this for FX / rewards)
    just_parried: bool = False
    just_guard_broke: bool = False
    just_parry_whiffed: bool = False
    # True if this shield press already scored a parry (shouldn't happen while still BLOCK)
    parry_connected: bool = False

    def clone(self) -> "Fighter":
        return Fighter(
            index=self.index,
            x=self.x,
            y=self.y,
            vx=self.vx,
            vy=self.vy,
            facing=self.facing,
            hp=self.hp,
            stamina=self.stamina,
            state=self.state,
            state_timer=self.state_timer,
            attack_timer=self.attack_timer,
            attack_def=self.attack_def,
            hit_connected=self.hit_connected,
            on_ground=self.on_ground,
            last_action=self.last_action,
            block_hold_frames=self.block_hold_frames,
            stamina_regen_cd=self.stamina_regen_cd,
            just_parried=self.just_parried,
            just_guard_broke=self.just_guard_broke,
            just_parry_whiffed=self.just_parry_whiffed,
            parry_connected=self.parry_connected,
        )

    @property
    def alive(self) -> bool:
        return self.hp > 0 and self.state != FighterState.KO

    @property
    def locked(self) -> bool:
        """Cannot freely choose movement/attacks."""
        return self.state in (
            FighterState.LIGHT,
            FighterState.HEAVY,
            FighterState.HITSTUN,
            FighterState.BLOCKSTUN,
            FighterState.GUARD_BREAK,
            FighterState.PARRY,
            FighterState.PARRY_WHIFF,
            FighterState.KO,
        )

    @property
    def is_attacking(self) -> bool:
        return self.state in (FighterState.LIGHT, FighterState.HEAVY)

    @property
    def is_shielding(self) -> bool:
        return self.state == FighterState.BLOCK and self.on_ground

    @property
    def in_parry_window(self) -> bool:
        """Fresh shield press: first PARRY_WINDOW_FRAMES of a hold."""
        return (
            self.is_shielding
            and 0 < self.block_hold_frames <= C.PARRY_WINDOW_FRAMES
        )

    @property
    def attack_active(self) -> bool:
        if not self.is_attacking or self.attack_def is None:
            return False
        ad = self.attack_def
        return ad.startup <= self.attack_timer < ad.startup + ad.active

    def body_rect(self) -> tuple[float, float, float, float]:
        half_w = C.FIGHTER_WIDTH * 0.5
        return (
            self.x - half_w,
            self.y - C.FIGHTER_HEIGHT,
            self.x + half_w,
            self.y,
        )

    def shield_rect(self) -> Optional[tuple[float, float, float, float]]:
        """Visual/logical shield plate in front of fighter."""
        if not self.is_shielding and self.state != FighterState.BLOCKSTUN:
            return None
        half_h = C.FIGHTER_HEIGHT * 0.55
        cx = self.x + self.facing * (C.FIGHTER_WIDTH * 0.55)
        top = self.y - C.FIGHTER_HEIGHT * 0.85
        w = 14.0
        if self.facing >= 0:
            return (cx, top, cx + w, top + half_h)
        return (cx - w, top, cx, top + half_h)

    def hitbox_rect(self) -> Optional[tuple[float, float, float, float]]:
        if not self.attack_active or self.attack_def is None:
            return None
        ad = self.attack_def
        if self.facing >= 0:
            left = self.x
            right = self.x + ad.range
        else:
            right = self.x
            left = self.x - ad.range
        top = self.y + ad.hitbox_y_offset - ad.hitbox_h * 0.5
        bottom = top + ad.hitbox_h
        return (left, top, right, bottom)

    def can_afford(self, cost: float) -> bool:
        return self.stamina >= cost

    def spend_stamina(self, amount: float) -> None:
        self.stamina = max(0.0, self.stamina - amount)
        self.stamina_regen_cd = C.STAMINA_REGEN_DELAY

    def apply_action(self, action: Action) -> None:
        """Begin processing an action for this frame (before physics)."""
        self.last_action = action
        self.just_parried = False
        self.just_guard_broke = False
        self.just_parry_whiffed = False

        if self.state == FighterState.KO:
            return

        # Locked recovery states
        if self.state in (
            FighterState.HITSTUN,
            FighterState.BLOCKSTUN,
            FighterState.GUARD_BREAK,
            FighterState.PARRY,
            FighterState.PARRY_WHIFF,
        ):
            return

        if self.is_attacking:
            return

        # --- Block / shield ---
        if action == Action.BLOCK and self.on_ground and self.stamina > 0:
            if self.state != FighterState.BLOCK:
                self.block_hold_frames = 0
                self.parry_connected = False
            self.state = FighterState.BLOCK
            self.block_hold_frames += 1
            self.vx = 0.0
            self.spend_stamina(C.BLOCK_HOLD_DRAIN)
            # Empty stamina while holding → drop shield (soft break)
            if self.stamina <= 0:
                self.stamina = 0.0
                self._release_block(whiff_check=True)
            return

        # Leaving block (release shield)
        if self.state == FighterState.BLOCK:
            self._release_block(whiff_check=True)
            # fall through so other actions on this frame can still apply
            if self.state == FighterState.PARRY_WHIFF:
                return

        if action == Action.LIGHT and self.on_ground:
            if self.can_afford(LIGHT_ATK.stamina_cost):
                self.spend_stamina(LIGHT_ATK.stamina_cost)
                self._start_attack(LIGHT_ATK, FighterState.LIGHT)
            return

        if action == Action.HEAVY and self.on_ground:
            if self.can_afford(HEAVY_ATK.stamina_cost):
                self.spend_stamina(HEAVY_ATK.stamina_cost)
                self._start_attack(HEAVY_ATK, FighterState.HEAVY)
            return

        if action == Action.JUMP and self.on_ground:
            if self.can_afford(C.JUMP_STAMINA_COST):
                self.spend_stamina(C.JUMP_STAMINA_COST)
                self.vy = C.JUMP_VELOCITY
                self.on_ground = False
                self.state = FighterState.JUMP
            return

        # Horizontal movement (costs stamina)
        moving = action in (Action.LEFT, Action.RIGHT)
        if moving:
            drain = (
                C.WALK_STAMINA_PER_FRAME
                if self.on_ground
                else C.AIR_MOVE_STAMINA_PER_FRAME
            )
            if self.stamina < C.MOVE_MIN_STAMINA:
                # Exhausted — can't walk
                if self.on_ground:
                    self.vx = 0.0
                    self.state = FighterState.IDLE
                return
            self.spend_stamina(drain)
            speed = C.WALK_SPEED if self.on_ground else C.WALK_SPEED * C.AIR_CONTROL
            if action == Action.LEFT:
                self.vx = -speed
            else:
                self.vx = speed
            if self.on_ground:
                self.state = FighterState.WALK
            return

        if self.on_ground:
            self.vx = 0.0
            self.state = FighterState.IDLE

    def _release_block(self, whiff_check: bool) -> None:
        """Drop shield. Short tap with no connect = parry whiff recovery."""
        held = self.block_hold_frames
        self.block_hold_frames = 0
        if (
            whiff_check
            and not self.parry_connected
            and 0 < held <= C.PARRY_WINDOW_FRAMES
        ):
            # Fished for a parry and got nothing — punish
            self.just_parry_whiffed = True
            self.state = FighterState.PARRY_WHIFF
            self.state_timer = C.PARRY_WHIFF_RECOVERY
            self.spend_stamina(C.PARRY_WHIFF_STAMINA_COST)
            self.vx = 0.0
        else:
            self.state = FighterState.IDLE
        self.parry_connected = False

    def _start_attack(self, attack: AttackDef, state: FighterState) -> None:
        self.state = state
        self.attack_def = attack
        self.attack_timer = 0
        self.hit_connected = False
        self.vx = 0.0
        self.block_hold_frames = 0

    def tick_timers_and_physics(self) -> None:
        """Advance timers, regen stamina, integrate motion."""
        self._regen_stamina()

        if self.state == FighterState.KO:
            self.vx *= 0.9
            self._integrate()
            return

        if self.state in (
            FighterState.HITSTUN,
            FighterState.BLOCKSTUN,
            FighterState.GUARD_BREAK,
            FighterState.PARRY,
            FighterState.PARRY_WHIFF,
        ):
            self.state_timer -= 1
            self.vx *= 0.85
            if self.state_timer <= 0:
                self.state = FighterState.IDLE if self.on_ground else FighterState.JUMP
                self.vx = 0.0
                self.block_hold_frames = 0
            self._integrate()
            return

        if self.is_attacking and self.attack_def is not None:
            self.attack_timer += 1
            if self.attack_timer >= self.attack_def.total_frames:
                self.state = FighterState.IDLE if self.on_ground else FighterState.JUMP
                self.attack_def = None
                self.attack_timer = 0
                self.hit_connected = False
            self._integrate()
            return

        if self.state == FighterState.BLOCK:
            self.vx = 0.0

        self._integrate()

        if self.on_ground and self.state == FighterState.JUMP:
            self.state = FighterState.IDLE

    def _regen_stamina(self) -> None:
        if self.stamina_regen_cd > 0:
            self.stamina_regen_cd -= 1
            return
        # No regen while attacking, blocking, walking hard, or broken
        if self.state in (
            FighterState.LIGHT,
            FighterState.HEAVY,
            FighterState.BLOCK,
            FighterState.BLOCKSTUN,
            FighterState.GUARD_BREAK,
            FighterState.PARRY_WHIFF,
            FighterState.WALK,
            FighterState.KO,
        ):
            return
        self.stamina = min(C.MAX_STAMINA, self.stamina + C.STAMINA_REGEN_PER_FRAME)

    def _integrate(self) -> None:
        if not self.on_ground or self.vy < 0:
            self.vy = min(self.vy + C.GRAVITY, C.MAX_FALL_SPEED)

        self.x += self.vx
        self.y += self.vy

        min_x = C.WALL_MARGIN
        max_x = C.STAGE_WIDTH - C.WALL_MARGIN
        if self.x < min_x:
            self.x = min_x
            self.vx = 0.0
        elif self.x > max_x:
            self.x = max_x
            self.vx = 0.0

        if self.y >= C.GROUND_Y:
            self.y = C.GROUND_Y
            self.vy = 0.0
            self.on_ground = True
        else:
            self.on_ground = False

    def facing_attacker(self, attacker_x: float) -> bool:
        """True if we are looking toward the attacker (required to shield/parry)."""
        if attacker_x >= self.x:
            return self.facing > 0
        return self.facing < 0

    def take_hit(
        self,
        damage: float,
        hitstun: int,
        knockback: float,
        direction: int,
        attacker_x: float,
        chip: float,
        block_stamina_cost: float,
        attack_blockstun: int,
    ) -> tuple[float, str]:
        """
        Apply a hit. Returns (damage_dealt, result) where result is one of:
        'hit' | 'block' | 'parry' | 'guard_break' | 'none'
        """
        if self.state == FighterState.KO:
            return 0.0, "none"

        # Parry success flash is invulnerable
        if self.state == FighterState.PARRY:
            return 0.0, "none"

        can_defend = (
            self.state == FighterState.BLOCK
            and self.on_ground
            and self.facing_attacker(attacker_x)
            and self.stamina > 0
        )

        if can_defend and self.in_parry_window:
            # Perfect parry: no damage taken; counter-damage applied to attacker in match
            self.just_parried = True
            self.parry_connected = True
            self.state = FighterState.PARRY
            self.state_timer = C.PARRY_FLASH_FRAMES
            self.block_hold_frames = 0
            self.attack_def = None
            self.attack_timer = 0
            self.stamina = min(C.MAX_STAMINA, self.stamina + C.PARRY_STAMINA_REFUND)
            self.vx = 0.0
            return 0.0, "parry"

        # Hit during parry-whiff recovery = free punish (full damage, already falls through)

        if can_defend:
            # Normal shield block
            self.spend_stamina(block_stamina_cost)
            if self.stamina <= 0:
                # Guard break — eat full damage + long stun
                self.stamina = 0.0
                self.just_guard_broke = True
                dealt = damage
                self.hp = max(0.0, self.hp - dealt)
                self.block_hold_frames = 0
                self.attack_def = None
                self.attack_timer = 0
                if self.hp <= 0:
                    self._ko()
                    self.vx = direction * knockback
                    return dealt, "guard_break"
                self.state = FighterState.GUARD_BREAK
                self.state_timer = C.GUARD_BREAK_STUN
                self.vx = direction * knockback * 1.1
                return dealt, "guard_break"

            dealt = chip
            self.hp = max(0.0, self.hp - dealt)
            self.state = FighterState.BLOCKSTUN
            self.state_timer = attack_blockstun
            self.block_hold_frames = 0
            self.vx = direction * knockback * C.BLOCK_PUSHBACK_SCALE
            if self.hp <= 0:
                self._ko()
            return dealt, "block"

        # Clean hit
        dealt = damage
        self.hp = max(0.0, self.hp - dealt)
        self.block_hold_frames = 0
        self.attack_def = None
        self.attack_timer = 0
        if self.hp <= 0:
            self._ko()
            self.vx = direction * knockback
            return dealt, "hit"

        self.state = FighterState.HITSTUN
        self.state_timer = hitstun
        self.vx = direction * knockback
        return dealt, "hit"

    def apply_parry_punish(self, direction: int, is_heavy: bool = False) -> float:
        """
        Called on attacker when their hit is parried.
        Deals counter damage and long stun. Returns damage dealt.
        """
        if self.state == FighterState.KO:
            return 0.0

        dmg = C.PARRY_COUNTER_DAMAGE + (
            C.PARRY_COUNTER_HEAVY_BONUS if is_heavy else 0.0
        )
        self.hp = max(0.0, self.hp - dmg)
        self.attack_def = None
        self.attack_timer = 0
        self.hit_connected = True
        self.block_hold_frames = 0
        self.vx = direction * C.PARRY_COUNTER_KNOCKBACK

        if self.hp <= 0:
            self._ko()
            return dmg

        self.state = FighterState.HITSTUN
        self.state_timer = C.PARRY_STUN_ATTACKER
        return dmg

    def _ko(self) -> None:
        self.hp = 0.0
        self.state = FighterState.KO
        self.state_timer = 0
        self.attack_def = None
        self.block_hold_frames = 0
        self.parry_connected = False
